Applications/ConsoleReferenceClient/UAClient.cs

```diff
@@ -124,19 +138,46 @@ public async Task<bool> ConnectAsync(string serverUrl, bool useSecurity = true)
                 }
                 else
                 {
-                    m_output.WriteLine("Connecting to... {0}", serverUrl);
+                    ITransportWaitingConnection connection = null;
+                    EndpointDescription endpointDescription = null;
+                    if (m_reverseConnectManager != null)
+                    {
+                        m_output.WriteLine("Waiting for reverse connection to.... {0}", serverUrl);
+                        do
+                        {
+                            using (var cts = new CancellationTokenSource(30_000))
+                            {
+                                connection = await m_reverseConnectManager.WaitForConnection(new Uri(serverUrl), null, cts.Token);
+                                if (connection == null)
+                                {
+                                    throw new ServiceResultException(StatusCodes.BadTimeout, "Waiting for a reverse connection timed out.");
+                                }
+                                if (endpointDescription == null)
+                                {
+                                    m_output.WriteLine("Discover reverse connection endpoints....");
+                                    endpointDescription = CoreClientUtils.SelectEndpoint(m_configuration, connection, useSecurity);
+                                    connection = null;
+                                }
+                            }
+                        } while (connection == null);
```

Comment:
Wont this loop forever when line 159 is setting connecton to null?

Reply:
should timeout, throw exception etc.

Reply:
But if Select returns a description then connection is set to null, without breaking, so the while will be true and loop again, no?

Reply:
the first channel is used by the discovery, the selected profile needs a new channel, so it will loop again but the select endpoint would be skipped. It loops only once.
A possible optimization to check, if the selected endpoint is security none the same channel could be used.